livre/core/pipeline/PortData.h

```diff
@@ -0,0 +1,93 @@
+/* Copyright (c) 2011-2016, EPFL/Blue Brain Project
+ *                     Ahmet Bilgili <ahmet.bilgili@epfl.ch>
+ *
+ * This file is part of Livre <https://github.com/BlueBrain/Livre>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef _PortData_h_
+#define _PortData_h_
+
+#include <livre/core/types.h>
+
+#include "PortType.h"
+
+namespace livre
+{
+
+/**
+ * PortData class is base class for keeping the track for types
+ * by using the std::type_index.
+ */
+class PortData : public PortType
+{
+protected:
+    explicit PortData( const std::type_index& dataType )
+        : PortType( dataType) {}
+    virtual ~PortData() {}
+};
+
+/**
+ * Holds the T typed data.
+ */
+template< class T>
+struct PortDataT final : public PortData
```

Comment:
You can do something similar to what you have, keeping the template class inside an implementation class, that's what I mean by "needs to use a pimpl".